### PR TITLE
xds: Implement server-side security

### PR DESCRIPTION
**Pull request by easwars**
Summary of changes:
- Add security integration to `xds.GRPCServer`:
  - Retrieve configured transport credentials from the underlying `grpc.Server`
  - Handle security configuration in received `Listener` resource
    - create appropriate certificate providers
    - setup `HandshakeInfo` to be passed to `xds.Credentials`
    - wrap the underlying `net.Conn`
      - support for setting and getting deadline
      - support for retrieving the `HandshakeInfo`
- e2e test which exercises xDS security with a file watcher plugin.
- A way to retrieve the transport credentials configured on a `grpc.Server` through the `internal` package.

### CODE REVIEW

**Comment by easwars** [comment]
@dfawley : This is now ready to be looked at. Thanks.

File: xds/server.go [hunk 1]
```diff
@@ -233,6 +251,116 @@ func (s *GRPCServer) newListenerWrapper(lis net.Listener) (*listenerWrapper, err
 	return lw, nil
 }
 
+// watchUpdate wraps the information received from a registered LDS watcher.
```
Comment (dfawley) [inline xds/server.go:254]: `listenerUpdate`?

File: xds/server.go [hunk 2]
```diff
@@ -233,6 +251,116 @@ func (s *GRPCServer) newListenerWrapper(lis net.Listener) (*listenerWrapper, err
...
+	// If we got all the way here, it means the received update was a good one.
+	update.goodUpdate.Fire()
+}
+
+func (s *GRPCServer) handleSecurityConfig(config *xdsclient.SecurityConfig, lw *listenerWrapper) error {
+	// If xdsCredentials are not in use, i.e, the user did not want to get
+	// security configuration from the control plane, we should not be acting on
+	// the received security config here. Doing so poses a security threat.
+	if !s.xdsCredsInUse {
+		return nil
```
Comment (dfawley) [inline xds/server.go:294]: Should we return an error?  Or at least log something?  (Except I guess this is expected if `config==nil`, right?)

File: xds/server.go [hunk 3]
```diff
@@ -233,6 +251,116 @@ func (s *GRPCServer) newListenerWrapper(lis net.Listener) (*listenerWrapper, err
...
+		CertName:     certName,
+		WantIdentity: wantIdentity,
+		WantRoot:     wantRoot,
+	})
+	if err != nil {
+		// This error is not expected since the bootstrap process parses the
+		// config and makes sure that it is acceptable to the plugin. Still, it
+		// is possible that the plugin parses the config successfully, but its
+		// Build() method errors out.
+		return nil, fmt.Errorf("xds: failed to get security plugin instance (%+v): %v", cfg, err)
```
Comment (dfawley) [inline xds/server.go:359]: Is this `xds` going to be redundant?

File: xds/server.go [hunk 4]
```diff
@@ -279,7 +407,16 @@ type listenerWrapper struct {
 	net.Listener
 	cancelWatch func()
 
-	// TODO(easwars): Add fields for certificate providers.
+	// A small race exists in the xdsClient code where an xDS response is
+	// received while the user is calling cancel(). In this small window the
+	// registered callback can be called after the watcher is canceled.
```
Comment (dfawley) [inline xds/server.go:412]: Why is this comment relevant to this field?

File: xds/server.go [hunk 5]
```diff
@@ -289,23 +426,68 @@ func (l *listenerWrapper) Accept() (net.Conn, error) {
...
 // based on security configuration received in the LDS response.
 func (l *listenerWrapper) Close() error {
+	l.closed.Fire()
 	l.Listener.Close()
 	if l.cancelWatch != nil {
 		l.cancelWatch()
 	}
+	if l.cachedIdentity != nil {
+		l.cachedIdentity.Close()
+	}
```
Comment (dfawley) [inline xds/server.go:443]: I think there's a race between this function and a security update happening.  One way to avoid that would be to access this field atomically; doing a `Load` here and a `Swap` when a new one is assigned due to an update.  Or add a mutex.

**Review by dfawley — CHANGES_REQUESTED** [review]

File: xds/server.go (hunk 3 above)
Comment (easwars) [reply xds/server.go:365]: The convention seems to be that we add the component prefix to errors that we generate. The same or closely related prefix might also get added by the prefix logger if the returned error is logged at some point. I still haven't looked at any real logs to see how useful/bad these prefixes are getting.

File: xds/server.go (hunk 4 above)
Comment (easwars) [reply xds/server.go]: I've clarified it a bit.
Basically the xdsClient code documents this race and says that callers need to handle it.

File: xds/server.go (hunk 5 above)
Comment (easwars) [reply xds/server.go]: Thanks. Added a mutex.

**Review by easwars — COMMENTED** [review]
Thanks for the review. PTAL.

File: xds/server.go (hunk 2 above)
Comment (dfawley) [reply xds/server.go:297]: Aha, sorry I saw "poses a security threat" and thought that meant we had a near-miss.

File: xds/server.go (hunk 3 above)
Comment (dfawley) [reply xds/server.go:365]: This isn't a convention we should have, *unless* the error is leaving the xds component.  Let's be careful about these so our logs are clear and not cluttered.

File: xds/server.go [hunk 6]
```diff
@@ -154,6 +167,18 @@ func (s *GRPCServer) Serve(lis net.Listener) error {
...
 		return err
 	}
+
+	// If xds credentials were specified by the user, but bootstrap configs do
+	// not contain any certificate provider configuration, it is better to fail
+	// right now rather than failing when attempting to create certificate
+	// providers after receiving an LDS response with security configuration.
+	if s.xdsCredsInUse {
+		bc := s.xdsC.BootstrapConfig()
+		if bc == nil || bc.CertProviderConfigs == nil {
```
[the remark this answers is not in the document]
Comment (dfawley) [reply xds/server.go]: You still need `bc == nil` or you will panic if it is.  Unless it's impossible to be `nil` -- is it?

File: xds/server.go [hunk 7]
```diff
@@ -338,13 +342,15 @@ func (s *GRPCServer) handleSecurityConfig(config *xdsclient.SecurityConfig, lw *
 	lw.xdsHI.SetRootCertProvider(rootProvider)
 	lw.xdsHI.SetIdentityCertProvider(identityProvider)
 	lw.xdsHI.SetRequireClientCert(config.RequireClientCert)
+	lw.providerMu.Unlock()
```
Comment (dfawley) [inline xds/server.go:345]: You can move this above the `xdsHI` parts, right?

File: xds/internal/test/xds_server_integration_test.go [hunk 8]
```diff
@@ -149,6 +327,99 @@ func (s) TestServerSideXDS(t *testing.T) {
...
+	defer cancel()
+	cc, err := grpc.DialContext(ctx, lis.Addr().String(), grpc.WithTransportCredentials(clientCreds))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+
+	// We don't set 'waitForReady` here since we want this call to failfast.
+	client := testpb.NewTestServiceClient(cc)
+	if _, err := client.EmptyCall(ctx, &testpb.Empty{}); err == nil {
```
Comment (dfawley) [inline xds/internal/test/xds_server_integration_test.go:402]: Stronger test: `status.Convert(err).Code() != codes.Unavailable`

**Review by dfawley — CHANGES_REQUESTED** [review]

File: xds/server.go (hunk 7 above)
Comment (easwars) [reply xds/server.go:345]: The `HandshakeInfo` type does use a mutex in its methods. In a very unlikely scenario, we could have two Listener updates come right after one another, and having the lock for the whole duration will make sure that the contents of `xdsHI` at the end of it all will be sane.

File: xds/internal/test/xds_server_integration_test.go (hunk 8 above)
Comment (easwars) [reply xds/internal/test/xds_server_integration_test.go]: Done.

File: xds/server.go (hunk 3 above)
Comment (easwars) [reply xds/server.go:365]: Can I leave it as-is for now? We should probably discuss this and come up with a convention that we can follow.

**Comment by easwars** [comment]
Had to rebase because of a merge conflict and the comments are not all over the place :(

**Comment by easwars** [comment]
For comments where I'm not able to reply inline:
- Added the `bc == nil` check back
- As for polluting the logs with our prefixes: can I leave it as-is for now. We should discuss this and come to a consensus on what would be the best strategy forward. It might require some wholesale changes
- As for the type in the comment in the `internal` package: `server.go` which implements `internal.GetServerCredentials` accepts an `interface{}` and not a `*grpc.Server`. It then goes ahead and type asserts to a `*grpc.Server`. I did this so that tests which fake out the underlying `grpc.Server` will continue to work without extensive changes. Let me know if you still want me to change the type in the comment. 

Thanks.

**Comment by dfawley** [comment]
> As for the type in the comment in the internal package: server.go which implements internal.GetServerCredentials accepts an interface{} and not a *grpc.Server. It then goes ahead and type asserts to a *grpc.Server. I did this so that tests which fake out the underlying grpc.Server will continue to work without extensive changes. Let me know if you still want me to change the type in the comment.

Which tests do this?

File: xds/server.go (hunk 3 above)
Comment (easwars) [reply xds/server.go:365]: I've removed the xds prefix from the errors returned from here.

I didn't mean to pushback on the change you are suggesting here. I'm only saying that this is widespread and that we should talk about it and decide how we want to go about structuring error messages.

**Review by easwars — COMMENTED** [review]
> > As for the type in the comment in the internal package: server.go which implements internal.GetServerCredentials accepts an interface{} and not a *grpc.Server. It then goes ahead and type asserts to a *grpc.Server. I did this so that tests which fake out the underlying grpc.Server will continue to work without extensive changes. Let me know if you still want me to change the type in the comment.
> 
> Which tests do this?

Tests in `xds/server_test.go` use a `fakeGRPCServer` to make sure that things like `Serve()`, `Stop()` etc on the `xds.GRPCServer` ends up calling the appropriate methods on the `grpc.Server`.

File: internal/internal.go [hunk 9]
```diff
@@ -64,7 +64,7 @@ var (
 	// GetServerCredentials returns the transport credentials configured on a
 	// gRPC server. An xDS-enabled server needs to know what type of credentials
 	// is configured on the underlying gRPC server. This is set by server.go.
-	GetServerCredentials interface{} // func (interface{}) credentials.TransportCredentials
```
[the remark this answers is not in the document]
Comment (easwars) [reply internal/internal.go]: Fixed it now.

**Review by dfawley — COMMENTED** [review]
> Tests in xds/server_test.go use a fakeGRPCServer to make sure that things like Serve(), Stop() etc on the xds.GRPCServer ends up calling the appropriate methods on the grpc.Server.

Okay, I get what's happening now, thanks.

How about if we do the type assertion in the xds code instead, since that's where the interface is used instead of a direct `*grpc.Server`?  It seems like a minor change (it is), but I think it's a little cleaner if the grpc package and its (semi-)exported functions don't have to do type assertions like this.

Also, conceptually, it's more appropriate to determine what to do if the server *isn't* a `*grpc.Server` closer to the code that knows what *else* it might be.  E.g., the fake could support this feature as well, but doing the type assertion in grpc makes that harder.

**Comment by easwars** [comment]
> > Tests in xds/server_test.go use a fakeGRPCServer to make sure that things like Serve(), Stop() etc on the xds.GRPCServer ends up calling the appropriate methods on the grpc.Server.
> 
> Okay, I get what's happening now, thanks.
> 
> How about if we do the type assertion in the xds code instead, since that's where the interface is used instead of a direct `*grpc.Server`? It seems like a minor change (it is), but I think it's a little cleaner if the grpc package and its (semi-)exported functions don't have to do type assertions like this.
> 
> Also, conceptually, it's more appropriate to determine what to do if the server _isn't_ a `*grpc.Server` closer to the code that knows what _else_ it might be. E.g., the fake could support this feature as well, but doing the type assertion in grpc makes that harder.

Makes sense. Moved the type assertion to the xds code. Thanks. It does look cleaner now.

File: xds/server.go [hunk 10]
```diff
@@ -119,10 +119,18 @@ func NewGRPCServer(opts ...grpc.ServerOption) *GRPCServer {
...
 	s.logger.Infof("Created xds.GRPCServer")
 
-	creds := grpcGetServerCreds(s.gs)
-	if xc, ok := creds.(interface{ UsesXDS() bool }); ok && xc.UsesXDS() {
-		s.xdsCredsInUse = true
+	// We type assert our underlying gRPC server to the real grpc.Server here
+	// before trying to retrieve the configured credentials. This approach
+	// avoids performing the same type assertion in the grpc package which
+	// provides the implementation for internal.GetServerCredentials, and allows
+	// us to use a fake gRPC server in tests.
```
Comment (dfawley) [inline xds/server.go:126]: For future consideration: moving this to a function would allow it to be stubbed for testing:
```go
var usesXDS = func(s *GRPCServer) bool { ... }
```
in tests:
```go
usesXDS = func(s *GRPCServer) bool { return s.gs.(*fakeGRPCServer).whatever }
```

**Review by dfawley — APPROVED** [review]

File: xds/server.go (hunk 10 above)
Comment (easwars) [reply xds/server.go:126]: Ack
